core/services/gateway/handler.go

```diff
@@ -0,0 +1,46 @@
+package gateway
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/smartcontractkit/chainlink/v2/core/services/job"
+)
+
+type UserCallback interface {
+	SendResponse(msg *Message, errCode ErrorCode, errMsg string) error
+}
+
+// Handler implements service-specific logic for managing messages from users and nodes.
+// There is one Handler object created for each DON.
+//
+// The lifecycle of a Handler object is as follows:
+//   - Init() call
+//   - Start() call
+//   - a series of HandleUserMessage/HandleNodeMessage calls, executed in parallel
+//     (Handler needs to guarantee thread safety)
+//   - Close() call
+type Handler interface {
+	job.ServiceCtx
+
+	Init(connMgr DONConnectionManager, donConfig *DONConfig) error
+
+	HandleUserMessage(ctx context.Context, msg *Message, cb UserCallback) error
+
+	HandleNodeMessage(ctx context.Context, msg *Message, nodeAddr string) error
+}
+
+type HandlerType = string
+
+const (
+	Dummy HandlerType = "dummy"
+)
+
+func NewHandler(handlerType HandlerType) (Handler, error) {
```

Comment:
This does not look elegant, including `Init` function. A common pattern is to have NewXyzHandler per implementation. This would eliminate the switch here and the Init() method.

Reply:
I'll remove Init() but I'll keep the factory function here for now. I liked the Init() because it shows what components the Handler has access to but you're right that it's not Go style.